New tests environments

thenets opened the pull request:
Doesn't exist yet a simple way to generated artifacts in a controlled environment.

This PR is a possible approach to have everything being builded for the current source code state.

### Benefits:
- Easy way to test any individual commit
- Easy way to integrate CI process for PR
- Fixed version for main dependencies, like golang and python

### TODO:
- [x] Add build artifacts command
- [x] Allow change golang version
- [x] Allow change python version
- [x] Add command to run all tests on current source code
- [x] Create documentation
- [x] Create GitHub Action workflow to build artifacts 

### How to use:

- Build artifacts
  - Install podman
  - Go to `tests/`
  - Run `make artifacts`
  - The artifacts will be generated at `tests/artifacts`

thenets committed:
Add build-requirements.txt

thenets committed:
New container image for testing

Switch to centos8 and allow "any" Python version

It's not all Python versions allowed. Should not conflict with centos8 python version.
The current Python is 3.6.8, so is not possible to use any other 3.6.x version.

Changed to a pre-built python package

Major tests dir structure refactor

first attempt to run all tests

Cleanup vagrant playbook

shanemcd commented on tests/container/base/Dockerfile:
```diff
@@ -0,0 +1,53 @@
+FROM fedora:34
```
We just switched away from using Fedora as the base image in a recent PR. It may be better to keep the tooling consistent, especially since Fedora moves so fast.

thenets replied:
Oh, cool. I'll switch to CentOS 8.

thenets replied:
Done :)

shanemcd commented on tests/container/test/run-ci.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+export PATH=/home/test-user/receptor:/home/test-user/go/bin:$PATH
```
Where is this script used?

thenets replied:
My mistake. I copied from GitHub Action workflow tests. That will be used when I finish the task called "Add command to run all tests on current source code".

thenets replied:
Done :)

kdelee commented on tests/container/Makefile:
```diff
@@ -0,0 +1,28 @@
+image: image-base
+
+image-base:
+	podman build -t $(IMAGE_TAG_BASE) -f ./base/Dockerfile ../..
```
s/Dockerfile/Containerfile/g


in the interest of being consistent with our other projects

thenets replied:
Done ^^

kdelee commented on tests/container/base/Dockerfile:
```diff
@@ -0,0 +1,61 @@
+# Install specific python version
+RUN set -x \
+    && cd /tmp \
+    && wget -q https://www.python.org/ftp/python/${PYTHON_VERSION}/Python-${PYTHON_VERSION}.tgz \
```
as discussed, we can install the dnf module instead of building from source
https://www.server-world.info/en/note?os=CentOS_8&p=python&f=4

```
dnf module -y install python38
```
will have to tweak a few other things to use the right python binary

thenets replied:
Done :)

thenets committed:
Update .gitignore

thenets commented on tests/environments/container/Containerfile:
```diff
@@ -0,0 +1,51 @@
+    && dnf module install -y ${PYTHON_PKG_NAME} \
+    && dnf clean all \
+    && alternatives --set python /usr/bin/python${PYTHON_VERSION} \
+    # Install specific golang version
+    && cd /tmp/ \
+    && wget -q https://golang.org/dl/go${GOLANG_VERSION}.linux-amd64.tar.gz \
+    && rm -rf /usr/local/go \
+    && tar -C /usr/local -xzf go${GOLANG_VERSION}.linux-amd64.tar.gz \
+    && rm -f go${GOLANG_VERSION}.linux-amd64.tar.gz \
+    && echo 'export PATH=${PATH}:/usr/local/go/bin' >> /etc/bashrc
```
I'll need help for Go installation on CentOS 8.

If I run the Receptor CLI tests using the golang installation method above, then I got this error message:

```shell
+ go test -v ./functional/cli/cli_test.go
# runtime/cgo
cgo: exec /missing-cc: fork/exec /missing-cc: no such file or directory
FAIL    command-line-arguments [build failed]
FAIL
```

How to solve:

If I install the golang package using `yum`, everything works fine, but I can't specify a golang version this way.

```shell
yum install -y golang
```

I need help to find a way to make the first method works fine or find some way to install any (maybe not any) golang versions via `yum`.

thenets replied:
Was decided to use the stable version from CentOS 8 repo.

thenets committed:
WIP receptor-tester.sh

Add new README files

Move tests README to functional tests dir

Change test image to use stable centos 8 golang

Added tests docs

thenets committed:
Testing GH Action build

Testing GH Action build v1

Testing GH Action build v2

Testing GH Action build v3

Testing GH Action build v4

Testing GH Action build v5

Testing GH Action build v6

thenets commented:
I thing the latest pipeline break for random reasons. Should be fixed and triggered again.

thenets commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,65 @@
+
+name: Build Artifact
+
+on: 
+  push:
+    branches:
+      - '*'
+    # tags:        
+    #   - v*
```
I think should be better restrict this to some branches. Maybe only `devel` for now, and maybe we can add more later.

thenets replied:
CI removed. Maybe can be added in another PR.

thenets commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,65 @@
+        cd tests/ && make artifacts-no-cache
+      if: matrix.os == 'ubuntu-latest'
+
+    # - name: Build <macOS>
+    #   run: |
+    #     go build -v .
+    #     mkdir dist
+    #     mv receptor ./dist/receptor-darwin-x86_64
+    #   if: matrix.os == 'macOS-latest'
+      
+    # - name: Build <windows>
+    #   run: |
+    #     go build -v .
+    #     mkdir dist
+    #     mv receptor.exe ./dist/receptor-windows-x86_64.exe
+    #   if: matrix.os == 'windows-latest'
```
I don't know if I should build artifacts for other OS too or not.

thenets replied:
CI removed. Maybe can be added in another PR.

shanemcd commented on tests/receptor-tester.sh:
```diff
@@ -0,0 +1,161 @@
+#!/bin/bash
```
I'm not a fan of the hand-rolled bash scripts. 

We should use a testing library, either Go or Python.

thenets replied:
But it's just a way to invoke commands. This is not `pytest` wrap or something like this, it's just a way to run the existent tests inside containers. Should I use `Python` CLI to invoke commands instead of `bash`, already designed for CLI invoking?

The `README.md` that I created maybe can give you a better idea why I created this script/tool.

The reason I created this is to avoid `Makefile`. `Makefile` are cool for a lot of stuff, but it lacks documentation. At least with this `bash` script I'm able to invoke regular commands like `Makefile` (which also uses `sh` or `bash`) with a documentation embedded.

Go applications implies a dedicated dir, build processes, and dependencies installed. Python also requires those stuff and even an "requirements.txt" file if I need to use some dependency to create a CLI.

I can recreate this using Go or Python, but the consequence will be a lot of complexity been introduced. I also can convert this to regular Makefile commands, but it lacks context and documentation. Should I do this anyway?

thenets committed:
Set podman bin as a variable

thenets committed:
Add support podman and docker on macos and linux

thenets committed:
build artifact CI removed

thenets committed:
New container images

thenets committed:
Delete vagrant files

thenets committed:
Remove /tests/ from .dockerignore. Quay bug.

thenets committed:
Fix container image tag

thenets committed:
Add podman and ansible-runner to dev image

thenets committed:
Remove .dockerignore

thenets committed:
Revert "Remove .dockerignore"

This reverts commit 7a47c42b5c969ca9f5b7e30e6d978208f2f57347.
